### DISCUSSION TITLE
How to execute "setuptools.Command" via pyproject.toml

### DISCUSSION THREAD

**Discussion by buhtz** [pypa/setuptools#4115] [Q&A]
It seems I still miss some basics.

I know about [`setuptools.Command`](https://setuptools.pypa.io/en/latest/userguide/extension.html#setuptools.Command) and that I can somehow(?!) implement customized build steps with it. But I don't know how to execute something like this

```
class MyCommand(setuptools.Command):
    def run():
         print('I am in!')
```

Where to put this code? setup.py? Somewhere else?
What should I write into pyproject.toml?

The setuptools docu do talk about setup.cfg. But I want to avoid that extra config file.

How does `python3 -m pip install .` know to execute the code of `MyCommand`?

I try to create, document and publish some [example projects](https://codeberg.org/buhtz/tech-demo-python-packaging/src/branch/main/03_i18n_helloworld) to illustrate such use cases and their solutions.

**Comment by abravalheri** [comment]
Hi @buhtz, I feel that some of the questions have already been discussed in the other [thread](https://github.com/pypa/setuptools/discussions/3912).

> What should I write into pyproject.toml?

By following the example configuration in `setup.py` in the other [thread](https://github.com/pypa/setuptools/discussions/3912) you should not have to add anything in `pyproject.toml`.


For the sake of completeness, bellow there are some highlights of the previous conversation (I am adding emphasis and commentary):

> ... when it comes to Python specific packaging (i.e. the procedure that uses pip, wheel/sdist files and usually PyPI), there are 3 ***distinctive*** phases as described above by John: (a) first the sdist file is built, (b) then the wheel file is built from the sdist and (c) then pip will install the items it finds inside the wheel.
>
> You can use setuptools to customise which stuff you want to add inside the sdist or inside the wheel (i.e. the archive file you use to distribute code and that you can upload to PyPI).
> 
> But setuptools is ***not involved in the process of installing*** things[1](https://github.com/pypa/setuptools/discussions/3912#user-content-fn-1-627dc69819c8e1483710406fee2b0cff).

I am emphasizing ***distinct*** because they are separated and isolated phases in a pipeline.
`pip` will not call any custom setuptools command when ***installing*** things.

When you call `pip install` with a directory, only then it will call setuptools to build a `.whl` file. At that moment in time you can use a sub-command to add files to this `.whl` archive (which is a glorified zip). If you use the configuration suggested in the other thread, the sub-command will be automatically called at that moment.
